To build ppp with dietlibc, we need TIOCSETD, which is defined only if USE_MISC is defined
Index: dietlibc-0.29/include/sys/mips-ioctl.h
===================================================================
--- dietlibc-0.29.orig/include/sys/mips-ioctl.h	2007-07-27 13:03:53.000000000 +0000
+++ dietlibc-0.29/include/sys/mips-ioctl.h	2007-07-27 13:04:22.000000000 +0000
@@ -1,3 +1,5 @@
+#define __USE_MISC
+
 #define _IOC_NRBITS	8
 #define _IOC_TYPEBITS	8
 #define _IOC_SIZEBITS	13
@@ -177,3 +179,4 @@
 #define SIOCGPGRP	_IOR('s', 9, pid_t)
 
 #define SIOCGSTAMP	0x8906			/* Get stamp - linux-specific */
+#undef __USE_MISC
